bloom-housing/bloom
# feat: migrate afs job off of redis

## Issue Overview

This PR addresses #3191 


## Description

The current setup for the application flagged sets cron job is never getting hit. This is due to its reliance on the bull nestjs library that uses redis. For an unknown reason the redis setup in heroku is connecting but nothing is getting set in it. To solve this problem this PR removes the dependence on redis and uses nestjs's scheduler functionality.

Things to note:
- The cron frequency can be set as a .env variable called `AFS_PROCESSING_CRON_STRING`
- Due to us using 3 dynos in our production environment if we were to use the nestjs cron jobs without any changes it would get run three times simultaneously and potentially crash. To solve this I have added a step where a random number between 0 and 59 is put at the beginning of the cron string. This makes it so it will run at a random second at the cron run time and theoretically the jobs will be run at a staggered interval. Also when the cron job is triggered a new timestamp is saved to the db to indicate the last run time of the job. Before running it checks to make sure that that the last run time is over 5 minutes old.
  - Sample example: Cron job is set to run at midnight everyday. dyno 1 randomly selects 24 as the seconds, dyno 2 selects 11, and dyno 3 selects 53.
    - dyno 2 triggers first at 12:00:11. It checks the db and last run was almost 24 hours ago.
    - dyno 2 updates the "last run time" to 12:00:11
    - dyno 2 runs the processing
    - dyno 1 triggers at 12:00:24. It checks the db and last run was 13 seconds ago so it doesn't run
    - dyno 3 triggers at 12:00:53. It checks the db and last run was 42 seconds ago so it doesn't run

## How Can This Be Tested/Reviewed?

The easiest way to test this is to add a new record to your .env file for backend core for `AFS_PROCESSING_CRON_STRING`. Sample `AFS_PROCESSING_CRON_STRING="0,10,30 * * * *"` (runs every hour at 0, 10, and 30 minutes). When you start up the application you should see logging of "running the Application flagged sets cron job" to appear within the minute you set in the environment variable. I'd recommend testing a couple of variations of the cron string. Note that you need to restart the application if you change the .env file.

The database table of `cron_job` should also be added with a new entry of name `AFS_CRON_JOB` after the cron job first gets triggered. Make sure the last run time is updated correctly

## Checklist:


## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.
